security/sensorhub-security-oauth/src/main/java/org/sensorhub/impl/security/oauth/OAuthAuthenticator.java

```diff
@@ -246,98 +247,121 @@ public Authentication validateRequest(ServletRequest req, ServletResponse resp,
-                    {
-                        // if ID token, assume it's a JWT token that contains needed info
-                        log.debug("ID Token = " + idToken);
-                        
-                        String[] chunks = idToken.split("\\.");
+
+                    if (clientCredentialsToken == null) {
+
+                        if (idToken != null)
+                        {
+                            // if ID token, assume it's a JWT token that contains needed info
+                            log.debug("ID Token = " + idToken);
+
+                            String[] chunks = idToken.split("\\.");
+                            Base64.Decoder decoder = Base64.getUrlDecoder();
+                            String header = new String(decoder.decode(chunks[0]));
+                            String payload = new String(decoder.decode(chunks[1]));
+                            log.debug("ID Token Header = " + header);
+                            log.debug("ID Token Payload = " + payload);
+                            userInfo = JsonParser.parseString(payload);
+                        }
+                        else
+                        {
+                            // if only an access token, we need to call the userinfo endpoint
+                            log.debug("OAuth Token = " + accessToken);
+
+                            // request user info
+                            OAuthClientRequest bearerClientRequest = new OAuthBearerClientRequest(config.userInfoEndpoint)
+                                    .setAccessToken(accessToken)
+                                    //.buildQueryMessage();
+                                    .buildHeaderMessage();
+                            OAuthResourceResponse resourceResponse = oAuthClient.resource(bearerClientRequest, HttpMethod.GET, OAuthResourceResponse.class);
+                            log.debug("UserInfo = " + resourceResponse.getBody());
+                            userInfo = JsonParser.parseString(resourceResponse.getBody());
+                        }
+
+                    } else {
+
+                        // if client credentials token, assume it's a JWT token that contains needed info
+                        log.debug("Client Credentials Token = " + clientCredentialsToken);
+
+                        String[] chunks = clientCredentialsToken.split("\\.");
+
                         Base64.Decoder decoder = Base64.getUrlDecoder();
                         String header = new String(decoder.decode(chunks[0]));
                         String payload = new String(decoder.decode(chunks[1]));
-                        log.debug("ID Token Header = " + header);
-                        log.debug("ID Token Payload = " + payload);
+                        log.debug("Client Credentials Token Header = " + header);
+                        log.debug("Client Token Payload = " + payload);
                         userInfo = JsonParser.parseString(payload);
```

Comment:
When a bearer token is received directly from a client, the signature needs to be verified and the signing authority must be authorized (e.g. be one of the authorized signers provided in the config). Otherwise this is completely unsecure as anybody can create a token to authenticate as an arbitrary user.

Reply:
I have made updates to get a JWKS from OIDC server and used to validate bearer token with audience and issuer. Implementation added a new config panel for bearer token properties to validate against when employing client credentials bearer token flow.  Used 0auth libraries with no current known vulnerabilities according to maven central.